File: 19_remove_nth_node_from_end_of_list.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        start = ListNode(0)
        start = ListNode(0)
        start.next = head
        p1 = start
        p2 = start

        for i in range(n + 1):
            if i == None:
                return None
            p1 = p1.next

        while p1:
            p1 = p1.next
            p2 = p2.next
        
        p2.next = p2.next.next
        return start.next

def convert_to_linked_list(input_list):
    head = None
    for i in range(len(input_list) - 1, -1, -1):
        new_head = ListNode(input_list[i], head)
        head = new_head
    return head

head = [1,2,3,4,5]
n = 2
print(Solution().removeNthFromEnd(convert_to_linked_list(head) , n))

File: linked_list/19_remove_nth_node_from_end_of_list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        // head = [1,2,3,4,5], n = 2 -> [1,2,3,5]
        ListNode res = new ListNode(0);
        res.next = head;
        int length = 0;
        while (head != null) {
            length++;
            head = head.next;
        }
        length -= n;
        
        ListNode first = res;
        while (length > 0) {
            length--;
            first = first.next;
        }
        first.next = first.next.next;
        return res.next;
        // O(N) + O(1)
    }
}